Error compiling WRF Plus in WRF 4.0.2 on module_wrf_top.f90

cross85 opened the pull request:
TYPE:  Bug fix 

KEYWORDS: WRF Plus, module_wrf_top.F, OpenMP error

SOURCE: Carlos Ross (DFM Consultores)

DESCRIPTION OF CHANGES: 
Problem: 
Getting an error when compiling WRF Plus with Intel compiler
```
../main/module_wrf_top.f90(756): error #6099: An ENDDO statement occurred without a corresponding DO or DO WHILE statement.
```
A DO loop in wrf_adtl_check_spot had a syntax error and a formulation error for one of the 
OpenMP parallel directives. 
1. A unnecessary trailing "&" on the last OpenMP directive caused the following  first line of the
    Fortran DO loop to not be recognized.
2. The OpenMP loop was missing closing directive, "!$OMP END PARALLEL DO",  after the DO loop.

Solution:
Remove the trailing "&" and add in the missing OpenMP directive after the DO loop.
I wasn't trying to compile with OpenMP, I actually chose 
```
option 10. (dmpar) INTEL (ifort/icc): Xeon (SNB with AVX mods) 
```
when I ran the configure. This error occurs only when I was building WRFPLUS.

Fixes #709

LIST OF MODIFIED FILES:  
M main/module_wrf_top.F

TESTS CONDUCTED: 
The code successfully compiled after including the modifications with Intel 19.0.1.

RELEASE NOTE: 
A syntax error was introduced with the v4.0 release that breaks the build of the WRFPLUS code (and only the WRFPLUS code) v19.0.1 of the Intel compiler. An OpenMP parallel loop was incorrectly constructed. That syntax error has been removed, though it is important to note that WRFPLUS never was intended to work with OpenMP, and the WRFPLUS code continues to only work with serial or DM build options.

davegill commented:
@cross85 @jjguerrette @liujake 
Carlos,
Great job on your PR! I made a few small changes, such as the commit message and to the base branch.

Would you add just a bit more information to the testing. For example, when would people see this problem? Is it only with a WRFPLUS build? Was OpenMP required for the build for this problem to show?

cross85 commented:
@cross85 @jjguerrette @liujake 
To replicate the problem, I used netcdf 4.6.2 (without netcdf4 or hdf5) and intel compilers.
and I follow these steps:

- ./configure wrfplus
- chose option 10 

> 10. (dmpar)   INTEL (ifort/icc): Xeon (SNB with AVX mods)

- changed the configure.wrf to this:

```
# Settings for    Linux x86_64 ppc64le i486 i586 i686, Xeon (SNB with AVX mods) ifort compiler with icc  (dmpar)
#

DESCRIPTION     =       INTEL ($SFC/$SCC): Xeon (SNB with AVX mods)
DMPARALLEL      =        1
OMPCPP          =       # -D_OPENMP
OMP             =       # -openmp -fpp -auto
OMPCC           =       # -openmp -fpp -auto
SFC             =       ifort
SCC             =       icc
CCOMP           =       icc
DM_FC           =       mpiifort
DM_CC           =       mpiicc -DMPI2_SUPPORT
FC              =       $(DM_FC)
CC              =       $(DM_CC) -DFSEEKO64_OK
LD              =       $(FC)
RWORDSIZE       =       8
PROMOTION       =       -r8        -real-size `expr 8 \* $(RWORDSIZE)` -i4
ARCH_LOCAL      =       -DNONSTANDARD_SYSTEM_FUNC -DCHUNK=64 -DXEON_OPTIMIZED_WSM5 -DOPTIMIZE_CFL_TEST  -DWRF_USE_CLM
OPTNOSIMD       =
OPTAVX          =       -xCORE-AVX512
CFLAGS_LOCAL    =       -w -O3 $(OPTAVX)
LDFLAGS_LOCAL   =       $(OPTAVX)
CPLUSPLUSLIB    =
ESMF_LDFLAG     =       $(CPLUSPLUSLIB)
FCOPTIM         =       -O3 $(OPTAVX)
FCREDUCEDOPT    =       $(FCOPTIM)
FCNOOPT         =       -O0 -fno-inline -no-ip
FCDEBUG         =       # -g $(FCNOOPT) -traceback # -fpe0 -check noarg_temp_created,bounds,format,output_conversion,pointers,uninit -ftrapuv -unroll0 -u
FORMAT_FIXED    =       -FI
FORMAT_FREE     =       -FR
FCSUFFIX        =
BYTESWAPIO      =       -convert big_endian
RECORDLENGTH    =       -assume byterecl
FCBASEOPTS_NO_G =       -w $(OMP) -auto -ftz -fno-alias -fp-model fast=1 -no-prec-div -no-prec-sqrt $(FORMAT_FREE) $(BYTESWAPIO) -auto -align array64byte #-vec-report6
FCBASEOPTS      =       $(FCBASEOPTS_NO_G) $(FCDEBUG)
MODULE_SRCH_FLAG =
TRADFLAG        =      -traditional-cpp
CPP             =      /lib/cpp -P -nostdinc
AR              =      ar
ARFLAGS         =      ru
M4              =      m4
RANLIB          =      ranlib
RLFLAGS         =
CC_TOOLS        =      gcc
```

- ./compile wrfplus
- After a while I get this error message:

```
mpiifort -o ../main/module_wrf_top.o -c -O3 -xCORE-AVX512 -w  -auto -ftz -fno-alias -fp-model fast=1 -no-prec-div -no-prec-sqrt -FR -convert big_endian -auto -align array64byte    -I../dyn_em -I../dyn_nmm  -I/home/cross/WRF/4.0.2_nc3/WRF_Plus/external/esmf_time_f90  -I/home/cross/WRF/4.0.2_nc3/WRF_Plus/main -I/home/cross/WRF/4.0.2_nc3/WRF_Plus/external/io_netcdf -I/home/cross/WRF/4.0.2_nc3/WRF_Plus/external/io_int -I/home/cross/WRF/4.0.2_nc3/WRF_Plus/frame -I/home/cross/WRF/4.0.2_nc3/WRF_Plus/share -I/home/cross/WRF/4.0.2_nc3/WRF_Plus/phys -I/home/cross/WRF/4.0.2_nc3/WRF_Plus/wrftladj -I/home/cross/WRF/4.0.2_nc3/WRF_Plus/chem -I/home/cross/WRF/4.0.2_nc3/WRF_Plus/inc -I/home/cross/Software/nc3/libs/netcdf/include  -r8        -real-size `expr 8 \* 8` -i4  ../main/module_wrf_top.f90
../main/module_wrf_top.f90(756): error #6099: An ENDDO statement occurred without a corresponding DO or DO WHILE statement.
   ENDDO
^
compilation aborted for ../main/module_wrf_top.f90 (code 1)
../configure.wrf:346: recipe for target '../main/module_wrf_top.o' failed
```

- Then I changed the code as discussed in #709, did the exact same steps as before and wrfplus was compiled without errors.

davegill commented:
@cross85 
Carlos,
We are close.
Would you provide the version of the Intel compiler please? 
Here's what I get on my machine:
```
> ifort --version
ifort (IFORT) 17.0.1 20161005
Copyright (C) 1985-2016 Intel Corporation.  All rights reserved.
```

davegill commented:
@jjguerrette @liujake 
Folks,
Would you please verify that this change does in fact not cause any troubles. We should be able to get bit-for-bit results with a before vs after comparison. When you guys report that, then this is good to commit.

liujake commented:
@jjguerrette Could you run a WRFDA regtest with this change? In the future, we should put WRFDA's (and WRF's) regtest into the cloud or somewhere so that all developers can run it themselves when making the changes.

davegill commented:
@liujake 
Jake,
That is a good idea.

cross85 commented:
@davegill 
Dave, I'm using this version:
```
> ifort --version
ifort (IFORT) 19.0.1.144 20181018
Copyright (C) 1985-2018 Intel Corporation.  All rights reserved.
```

cross85 commented:
I think it's just WRFPLUS that compiles this part of the code. I didn't have a problem before with WRF or WRFDA.

jjguerrette commented:
These code fixes look correct to me (as the person who wrote this section originally), and should be merged.  However, I know without running the WRFDA REGTEST that there will be no impact for two reasons:

1) The wrf_adtl_check_spot subroutine is never called during the REGTEST.  wrf_adtl_check_sum is used during the WTF suite tests, but not during the WRFDA REGTEST.

2) The only lines changed in the PR contain OMP directives that our compilers on cheyenne see as comments when OpenMP is disabled.  OpenMP is always disabled during the WRGFDA regression test.

The only way for OpenMP to become enabled is when the "-openmp" flag is provided to the Intel compiler, unless there is some strange behavior with Intel 19.0.  Carlos, when you say "changed configure.wrf to this", are you manually changing some of the settings?  If so, did you perform a "clean -a" between WRF and WRFPLUS compilation?  Does the "-openmp" flag show up in your compile log file?  Alternatively, does a line containing "use omp_lib" appear in your preprocessed module_wrf_top.f90 file?

jjguerrette commented:
In short, there has to be a specific reason we do not experience this error in our own build procedure.  The reason seems to be related to OpenMP even if it appears to be disabled in configure.wrf.

liujake commented:
I agree it should have no impact. I have no problem to not run WRFDA Regtest as we never turn on OpenMP compilation for WRFPlus and WRFDA.

cross85 commented:
@jjguerrette 
I changed the following  settings:

- to use Intel MPI:

```
DM_FC           =       mpiifort
DM_CC           =       mpiicc -DMPI2_SUPPORT
```

- to use AVX-512

```
OPTAVX          =       -xCORE-AVX512
```

The original one was like this:

```
DM_FC           =       mpif90 -f90=$(SFC)
DM_CC           =       mpicc -cc=$(SCC)

OPTAVX          =       -xAVX
```
In respect of the -openmp flag and "use omp_lib" doesn't appear on the compile log or the module_wrf_top.f90 file.
In any case, the -openmp flag shouldn't work, because since version 18 of intel compilers is changed to -qopenmp.

Here I attached the module_wrf_top.f90 that was created, in case you want to check it.
[module_wrf_top.zip](https://github.com/wrf-model/WRF/files/2626132/module_wrf_top.zip)

jjguerrette commented:
@cross85, thank you for checking.  I will have to settle with confusion this time around.

davegill commented:
@jjguerrette 
JJ,
Perhaps we would have had troubles when we switched to ifort 19.
